File: 算法/代码随想录算法PDF/00调试源代码/71.简化路径.py
#
# @lc app=leetcode.cn id=71 lang=python3
#
# [71] 简化路径
#

# @lc code=start
class Solution:
    def simplifyPath(self, path: str) -> str:
        st = []
        i = 0  # 遍历下标
        while i < len(path):
            if path[i] == '/':
                i += 1
            else:
                folder = ""
                while i < len(path) and path[i] != '/':
                    folder += path[i]
                    i += 1
                if folder == ".." and st:
                    st.pop()
                elif folder != ".." and folder != ".":
                    st.append(folder)
        res = "/"
        for f in st:
            res += f + "/"
        if len(res) > 1:
            res = res[:-1]
        return res

# @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/71.简化路径.cpp
/*
 * @lc app=leetcode.cn id=71 lang=cpp
 *
 * [71] 简化路径
 */

// @lc code=start
class Solution {
public:
    string simplifyPath(string path) {
        deque<string> st;
        for (int i = 0; i < path.size();) {
            if (path[i] == '/') {  // 遇到分隔符，继续向后遍历
                i++;
            } else {
                string folder;  // 文件夹名
                while (i < path.size() && path[i] != '/') {
                    folder += path[i++];
                }
                if (folder == ".." && !st.empty()) {  // 返回上一级
                    st.pop_back();
                } else if (folder != ".." && folder != ".") {  // 正常文件目录
                    st.push_back(folder);
                }
            }
        }
        string res = "/";
        for (string& s : st) {
            res += s + "/";
        }
        if (res.size() > 1) {
            res.pop_back();
        }
        return res;
    }
};
// @lc code=end

